services/vision/mlvision/ml_model.go

```diff
@@ -0,0 +1,227 @@
+	"strconv"
+	"strings"
+
+	"github.com/edaniels/golog"
+	"github.com/montanaflynn/stats"
+	"github.com/pkg/errors"
+	"go.opencensus.io/trace"
+
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/robot"
+	"go.viam.com/rdk/services/mlmodel"
+	"go.viam.com/rdk/services/vision"
+	"go.viam.com/rdk/utils"
+)
+
+var model = resource.NewDefaultModel("mlmodel")
+
+const (
+	// UInt8 is one of the possible input/output types for tensors.
+	UInt8 = "uint8"
+	// Float32 is one of the possible input/output types for tensors.
+	Float32 = "float32"
+)
+
+func init() {
+	resource.RegisterService(vision.Subtype, model, resource.Registration[vision.Service, *MLModelConfig]{
+		DeprecatedRobotConstructor: func(ctx context.Context, r any, c resource.Config, logger golog.Logger) (vision.Service, error) {
+			attrs, err := resource.NativeConfig[*MLModelConfig](c)
+			if err != nil {
+				return nil, err
+			}
+			actualR, err := utils.AssertType[robot.Robot](r)
+			if err != nil {
+				return nil, err
+			}
+			return registerMLModelVisionService(ctx, c.ResourceName().Name, attrs, actualR, logger)
+		},
+	})
+}
+
+// MLModelConfig specifies the parameters needed to turn an ML model into a vision Model.
+type MLModelConfig struct {
+	resource.TriviallyValidateConfig
+	ModelName string `json:"mlmodel_name"`
+}
+
+func registerMLModelVisionService(
+	ctx context.Context,
+	name string,
+	params *MLModelConfig,
+	r robot.Robot,
+	logger golog.Logger,
+) (vision.Service, error) {
+	_, span := trace.StartSpan(ctx, "service::vision::registerMLModelVisionService")
+	defer span.End()
+
+	mlm, err := mlmodel.FromRobot(r, params.ModelName)
+	if err != nil {
+		return nil, err
+	}
+	classifierFunc, err := attemptToBuildClassifier(mlm)
+	if err != nil {
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a classifier", params.ModelName))
+	} else {
+		logger.Infof("model %q fulfills a vision service classifier", params.ModelName)
+	}
+	detectorFunc, err := attemptToBuildDetector(mlm)
+	if err != nil {
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a detector", params.ModelName))
+	} else {
+		logger.Infof("model %q fulfills a vision service detector", params.ModelName)
+	}
+	segmenter3DFunc, err := attemptToBuild3DSegmenter(mlm)
+	if err != nil {
+		logger.Infof("%v", errors.Wrapf(err, "was not able to turn ml model %q into a 3D segmenter", params.ModelName))
+	} else {
+		logger.Infof("model %q fulfills a vision service 3D segmenter", params.ModelName)
+	}
+	// Don't return a close function, because you don't want to close the underlying ML service
+	return vision.NewService(name, r, nil, classifierFunc, detectorFunc, segmenter3DFunc)
+}
+
+// Unpack output based on expected type and force it into a []float64.
+func unpack(inMap map[string]interface{}, name string, md mlmodel.MLMetadata) []float64 {
+	var out []float64
+	me := inMap[name]
+	switch getTensorTypeFromName(name, md) {
+	case UInt8:
+		temp := me.([]uint8)
```

Comment:
it seems like you might as well do `switch v:= getTensorTypeFromName(name, md).()type {`!

Reply:
I might be misunderstanding but that method is gonna return a string every time. The string might say "uint8" or whatever, but it'll still be a string. I use consts for UInt8 and Float32 because they're the only types that matter right now.

Reply:
but we immediately cast to the type, so why not check the type?